src/SonarScanner.MSBuild.Shim/AdditionalFilesService.cs

```diff
@@ -74,33 +75,28 @@ public AdditionalFiles AdditionalFiles(AnalysisConfig analysisConfig, DirectoryI
             : PartitionAdditionalFiles(allFiles, analysisConfig);
     }
 
-    private List<string> GetAllFiles(IEnumerable<string> extensions, DirectoryInfo projectBaseDir)
-    {
-        var pattern = string.Join("|", extensions.Select(x => x.TrimStart('.') + "$").Distinct());
-        var regex = new Regex(pattern, RegexOptions.Compiled | RegexOptions.IgnoreCase, RegexTimeout);
-        return directoryWrapper
+    private List<FileInfo> GetAllFiles(IEnumerable<string> extensions, DirectoryInfo projectBaseDir) =>
+        directoryWrapper
             .EnumerateFiles(projectBaseDir.FullName, "*", SearchOption.AllDirectories)
-            .Where(regex.SafeIsMatch)
+            .Select(x => new FileInfo(x))
+            .Where(x => extensions.Any(e => x.Name.EndsWith(e, StringComparison.OrdinalIgnoreCase) && !x.Name.Equals(e, StringComparison.OrdinalIgnoreCase)))
             .ToList();
-    }
 
     private static bool HasUserSpecifiedSonarTests(AnalysisConfig analysisConfig) =>
         analysisConfig.LocalSettings.Exists(x => x.Id == SonarProperties.Tests);
 
-    private static AdditionalFiles PartitionAdditionalFiles(List<string> allFiles, AnalysisConfig analysisConfig)
+    private static AdditionalFiles PartitionAdditionalFiles(List<FileInfo> allFiles, AnalysisConfig analysisConfig)
     {
         var testExtensions = GetTestExtensions(analysisConfig.ServerSettings);
         if (testExtensions.Length == 0)
         {
             return new(allFiles, []);
         }
-        var testsPattern = string.Join("|", testExtensions);
-        var testsRegex = new Regex(testsPattern, RegexOptions.Compiled | RegexOptions.IgnoreCase, RegexTimeout);
-        var sources = new List<string>();
-        var tests = new List<string>();
+        var sources = new List<FileInfo>();
+        var tests = new List<FileInfo>();
         foreach (var file in allFiles)
         {
-            if (testsRegex.SafeIsMatch(file))
+            if (Array.Exists(testExtensions, x => file.Name.EndsWith(x) && !file.Name.Equals(x)))
```

Comment:
StringComparison for both string checks is missing. Please also add tests where the casing is relevant.